tracer/src/Datadog.Trace/Debugger/Expressions/ProbeProcessor.cs

```diff
@@ -0,0 +1,268 @@
+// <copyright file="ProbeProcessor.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using Datadog.Trace.Debugger.Configurations.Models;
+using Datadog.Trace.Debugger.RateLimiting;
+using Datadog.Trace.Debugger.Snapshots;
+using Datadog.Trace.Logging;
+using SnapshotProbe = Datadog.Trace.Debugger.Configurations.Models.SnapshotProbe;
+
+namespace Datadog.Trace.Debugger.Expressions
+{
+    internal class ProbeProcessor
+    {
+        protected static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(ProbeProcessor));
+
+        private ProbeExpressionEvaluator _evaluator;
+
+        public ProbeProcessor(ProbeDefinition probe)
+        {
+            var location = probe.Where.MethodName != null
+                               ? ProbeLocation.Method
+                               : ProbeLocation.Line;
+
+            var probeType = probe switch
+            {
+                SnapshotProbe { Capture: { } } => ProbeType.Snapshot,
+                MetricProbe => ProbeType.Metric,
+                _ => ProbeType.Log
+            };
+
+            ProbeInfo = new ProbeInfo(
+                probe.Id,
+                probeType,
+                location,
+                probe.EvaluateAt,
+                Templates: probe.Segments,
+                Condition: (probe as SnapshotProbe)?.When,
+                Metric: (probe as MetricProbe)?.Value);
+        }
+
+        internal ProbeInfo ProbeInfo { get; }
+
+        private ProbeExpressionEvaluator GetOrCreateEvaluator(MethodScopeMembers scopeMembers)
+        {
+            Interlocked.CompareExchange(ref _evaluator, new ProbeExpressionEvaluator(ProbeInfo.Templates, ProbeInfo.Condition, ProbeInfo.Metric, scopeMembers), null);
+            return _evaluator;
+        }
```

Comment:
I don't get it. `MethodScopeMembers` hold runtime data collected during method execution, and there is one `ProbeProcessor` instance per probe. Is creating the evaluator using the particular `MethodScopeMembers` is a bug? How will it work if there are two concurrent calls to a method with an instrumentation - how they will be distinguished?

From reading the code it seems you are creating an evaluator on method exit, but this evaluator lives in `ProbeProcessor`, and `ProbeProcessor` has an affinity to Probe Id.
The execution context should be totally separated from the probe processing functionality.